src/app/util/ember-compatibility-functions.cpp

```diff
@@ -575,36 +556,66 @@ CHIP_ERROR WriteSingleClusterData(ClusterInfo & aClusterInfo, TLV::TLVReader & a
     {
-        return apWriteHandler->AddStatus(attributePathParams, Protocols::InteractionModel::Status::UnsupportedAttribute);
-    }
+        AttributeAccessInterface * attrOverride = findAttributeAccessOverride(aClusterInfo.mEndpointId, aClusterInfo.mClusterId);
+        if (attrOverride != nullptr)
+        {
+            AttributeValueDecoder valueDecoder(aReader, apWriteHandler->GetAccessingFabricIndex());
+            ReturnErrorOnFailure(attrOverride->Write(aPath, valueDecoder));
 
-    if (attributeMetadata->IsReadOnly())
-    {
-        return apWriteHandler->AddStatus(attributePathParams, Protocols::InteractionModel::Status::UnsupportedWrite);
+            if (valueDecoder.TriedDecode())
+            {
+                status = Protocols::InteractionModel::Status::Success;
+                ExitNow();
+            }
+        }
     }
 
-    AttributeAccessInterface * attrOverride = findAttributeAccessOverride(aClusterInfo.mEndpointId, aClusterInfo.mClusterId);
-    if (attrOverride != nullptr)
     {
-        AttributeValueDecoder valueDecoder(aReader, apWriteHandler->GetAccessingFabricIndex());
-        ReturnErrorOnFailure(attrOverride->Write(aPath, valueDecoder));
+        const EmberAfAttributeMetadata * metadata =
+            emberAfLocateAttributeMetadata(aPath.mEndpointId, aPath.mClusterId, aPath.mAttributeId, CLUSTER_MASK_SERVER, 0);
 
-        if (valueDecoder.TriedDecode())
+        if (metadata == nullptr)
         {
-            return apWriteHandler->AddStatus(attributePathParams, Protocols::InteractionModel::Status::Success);
+            status = Protocols::InteractionModel::Status::UnsupportedAttribute;
+            ExitNow();
         }
+
+        if (metadata->IsReadOnly())
+        {
+            status = Protocols::InteractionModel::Status::UnsupportedWrite;
+            ExitNow();
+        }
+
+        CHIP_ERROR preparationError = CHIP_NO_ERROR;
+        uint16_t dataLen            = 0;
+        if ((preparationError = prepareWriteData(metadata, aReader, dataLen)) != CHIP_NO_ERROR)
+        {
+            ChipLogDetail(Zcl, "Failed to prepare data to write: %s", ErrorStr(preparationError));
+            status = Protocols::InteractionModel::Status::InvalidValue;
+            ExitNow();
+        }
+
+        if (dataLen > metadata->size)
+        {
+            ChipLogDetail(Zcl, "Data to write exceedes the attribute size claimed.");
+            status = Protocols::InteractionModel::Status::InvalidValue;
+            ExitNow();
+        }
+
+        status =
+            ToInteractionModelStatus(emberAfWriteAttributeExternal(aPath.mEndpointId, aPath.mClusterId, aPath.mAttributeId,
+                                                                   CLUSTER_MASK_SERVER, 0, attributeData, metadata->attributeType));
     }
 
-    auto imCode = WriteSingleClusterDataInternal(aPath, attributeMetadata, aReader, apWriteHandler);
-    return apWriteHandler->AddStatus(attributePathParams, imCode);
+exit:
```

Comment:
I was trying to figure out whether we could make use of `MakeDefer` here to avoid the indentation and careful exit orchestration, but if we want to return the AddStatus retval I guess that's hard...

Reply:
OK I redid the refactoring on top of master branch, and used return of that line `apWriteHandler->AddStatus(attributePathParams, status);` to make it simpler.